chore: improve gRPC overview page (#699)
vegaprotocol/documentation: docs/api/grpc/overview.md

sidebar_label: Using gRPC
hide_title: false
description: Learn how to use the gPRC API.
vega_network: TESTNET
---
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

gRPC is a low-latency, highly efficient communication method that allows a client and a server to send data over HTTP. Vega supports a gRPC API for interacting with the network and querying for and filtering data. 

## gRPC API and Protocol Buffers
gRPC uses Protocol Buffers (protobufs) as its underlying data definition. Protobuf messages are language-neutral data definitions that are used to generate language-specific data structures. Code generation of a protobuf message also generates functions to serialise each message so that they can be sent through gRPC to each RPC service endpoint.

This means that Vega's gRPC API is fully defined by its protobuf definitions, and so the protobuf definitions themselves provide complete documentation.

## Relationship to REST API
The buf ecosystem contains plugins that also allow automatic generation of REST API endpoints from the protobuf definitions. The result is that Vega's gRPC and REST API match exactly in both structure and functionality.

> As REST is a more familiar way to interact with a product than gRPC, experiencing Vega first through the REST API and using the REST documentation may help ease an initial integration, without any loss of functionality. Migrating later to using gRPC is then trivial since the mapping between REST and gRPC is 1-to-1.

Reviewer: One thing unclear to me (mostly because I don't know how the gRPC plugins work) is how the REST  endpoints are generated from the protobuf definitions - I don't think I would be able to make the connection between specific gRPC messages and REST endpoints without the REST docs.

Is this saying I should find it trivial to work-out what a REST endpoint is just from the proto definitions?

Author: I think the idea is that most people would go the other way around -- they'll play with the REST API first because its more familiar, and then work out how to go from the REST docs to gRPC. So in that situation they'll be looking at, say, the docs for this REST endpoint:
https://docs.vega.xyz/mainnet/api/rest/data-v2/trading-data-service-get-order
which has the heading "Get Order", wonder how to do it in gRPC and start tapping `.GetO` into their IDE which would autocomplete into `.GetOrder()` and they go "wow that was easy, I don't even _need_ to look at the gRPC docs"

If you were to want to go the other way and go from a gRPC call to the REST URL for it, the true answer is that the map is in this file which we could consider:
https://github.com/vegaprotocol/vega/blob/develop/protos/sources/data-node/grpc-rest-bindings.yml
but you could also just look through in the REST API docs to find it and the way they are grouped in the sidebar should help.

@candida-d do you think we should add a link to https://github.com/vegaprotocol/vega/blob/develop/protos/sources/data-node/grpc-rest-bindings.yml ? I'm worried it'll just make it seem more confusing.

Author: I've added two snippets showing how similar the using REST and gRPC are (and linked to the yml file that maps the url <-> gRPC function call), which is hopefully is more useful

```

</TabItem>

<TabItem value="go" label="Golang">

```go
package main

import (
	"context"
	"fmt"

	datanode "code.vegaprotocol.io/vega/protos/data-node/api/v2"
	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/credentials/insecure"
	"google.golang.org/grpc/status"
)

func main() {
	// set up a grpc connection
	grpcAddress := ""
	conn, err := grpc.Dial(grpcAddress, grpc.WithTransportCredentials(insecure.NewCredentials()))
	if err != nil {
		fmt.Println("unable to connect to data node", err)
		return
	}
	tradingDataClient := datanode.NewTradingDataServiceClient(conn)

	// fill in a request to get an order by its ID
	orderID := "01c25933750f9c3e35f38da9ee65c8b3eda165e914e86cad743b9effe826f2dc"
	request := &datanode.GetOrderRequest{OrderId: orderID}

	// call into the data node using the GetOrder endpoint
	response, err := tradingDataClient.GetOrder(context.Background(), request)
	switch status.Code(err) {
	case codes.OK:
		fmt.Println("found order:", response.Order)
		return
	case codes.NotFound:
		fmt.Println("order not found")
		return
	default:
		fmt.Println("unable to get order from a data node:", err)
	}

> }

Author: Yea fair enough, I've just going to remove these lines though because I don't think they are reachable:
```
case codes.OK:
    log.Fatal("found order:", response.Order)
```

sidebar_label: Using gRPC
hide_title: false
description: Learn how to use the gPRC API.
vega_network: TESTNET
---
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

gRPC is a low-latency, highly efficient communication method that allows a client and a server to send data over HTTP. Vega supports a gRPC API for interacting with the network and querying for and filtering data. 

## gRPC API and Protocol Buffers
gRPC uses Protocol Buffers (protobufs) as its underlying data definition. Protobuf messages are language-neutral data definitions that are used to generate language-specific data structures. Code generation of a protobuf message also generates functions to serialise each message so that they can be sent through gRPC to each RPC service endpoint.

This means that Vega's gRPC API is fully defined by its protobuf definitions, and so the protobuf definitions themselves provide complete documentation.

## Relationship to REST API
The buf ecosystem contains plugins that also allow automatic generation of REST API endpoints from the protobuf definitions. The result is that Vega's gRPC and REST API match exactly in both structure and functionality.

> As REST is a more familiar way to interact with a product than gRPC, experiencing Vega first through the REST API and using the REST documentation may help ease an initial integration, without any loss of functionality. Migrating later to using gRPC is then trivial since the mapping between REST and gRPC is 1-to-1.

Reviewer 2: ```suggestion
If you are new to interacting with gRPC, using REST may be a more familiar way to interact with the APIs. Experiencing Vega first through the REST API and using the REST documentation may help ease an initial integration, without any loss of functionality. Migrating later to using gRPC is then trivial since the mapping between REST and gRPC is 1-to-1.
```

  option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {tags: "Orders"};
}

// Request that is sent when executing the query for getting a single order
message GetOrderRequest {
  // Order ID to retrieve order information for.
  string order_id = 1 [(google.api.field_behavior) = REQUIRED];
  // Historic version number of the order to return. If not set, the most current version will be returned.
  optional int32 version = 2;
}

// Response received from the query for getting a single order
message GetOrderResponse {
  // Order details, if one was found.
  vega.Order order = 1;
}
```

The above protobuf definitions can be used to generate client and server code in a given language. This can be done by using the protocol-compiler `protoc` directly and supplier a output language, or by using CLI tools from the buf-ecosystem, namely `buf generate`. As an example, YAML configuration for compiling Vega's protobuf definition into Golang using `buf genreate` can be found here [here](https://github.com/vegaprotocol/vega/blob/develop/buf.gen.yaml)

Once generated they can imported and used from the generated language specific client stubs as follows:

<Tabs>

> <TabItem value="py" label="Python">

Author: I've done python and golang examples. I don't know if we know what other languages might be useful given our knowledge of people who have already integrated?

Reviewer 2: @wwestgarth Seems like this will do for now, and we can add more as we get requests/more data

## gRPC API and Protocol Buffers
gRPC uses Protocol Buffers (protobufs) as its underlying data definition. Protobuf messages are language-neutral data definitions that are used to generate language-specific data structures. Code generation of a protobuf message also generates functions to serialise each message so that they can be sent through gRPC to each RPC service endpoint.

This means that Vega's gRPC API is fully defined by its protobuf definitions, and so the protobuf definitions themselves provide complete documentation.

## Relationship to REST API
The buf ecosystem contains plugins that also allow automatic generation of REST API endpoints from the protobuf definitions. The result is that Vega's gRPC and REST API match exactly in both structure and functionality.

If you are new to interacting with gRPC, using REST may be a more familiar way to interact with the APIs. Experiencing Vega first through the REST API and using the REST documentation may help ease an initial integration, without any loss of functionality. Migrating later to using gRPC will then require minimal changes since the the input parameters and responses are contain the same data.

> As an example to show the similarities, below are two Python snippets of how to list transfers using both REST and gRPC:

Reviewer 2: ```suggestion
As an example to show the similarities, below are two Python snippets for how to list transfers using both REST and gRPC:
```
